### DISCUSSION TITLE
Can I Use WebSocket with Grapevine?

### DISCUSSION THREAD

**Discussion by scottoffen** [scottoffen/grapevine#67] [Q&A]
User @seikosantana asked about [WebSocket support in Grapevine 5](https://github.com/scottoffen/grapevine/issues/66). 

The short answer is I _think_ so. I'm not well versed on WebSocket, so there might be a better way to implement this, but here is the approach from what I understand about WebSocket.

In your route, you will need to cast the `IHttpContext` argument to the concrete class `HttpContext`, and you will find on that class the `Advanced` property. This property is the underlying `HttpListenerContext` for this request, which means you can access the [`AcceptWebSocketAsync`](https://docs.microsoft.com/en-us/dotnet/api/system.net.httplistenercontext.acceptwebsocketasync?view=net-5.0) method.

```csharp
[RestRoute("Get", "/path", Name = "Web Socket Example", Description = "Sample route for WebSocket")]
public async Task WebSocketHandler(IHttpContext context)
{
    if (context is HttpContext ctx)
    {
        if (ctx.Advanced.Request.IsWebSocketRequest)
        {
            var webSocketContext = await ctx.Advanced.AcceptWebSocketAsync("");
            // do your stuff here, I guess?
        }
    }

    await context.Response.SendResponseAsync(HttpStatusCode.Ok);
}
```

I'm open to suggestions on how to improve the experience here, but unless I can get some rather quick and concrete feedback, this is how it will be for the v5 release.

**Comment by seikosantana** [comment]
Hi,

Thank you for your response.
I havent checked github for a while. That snippet seems cool. I will try that out.

**Reply by scottoffen** [reply]
I haven't actually tested it yet, I've been on the road the past week. It might need some changes.

**Reply by seikosantana** [reply]
I have started to work on the websocket part with Grapevine and empty string can't be used as `subprotocol`. And I have not yet succeeded attempting connection from Postman
![image](https://user-images.githubusercontent.com/35630407/174532706-4835c8e5-196a-41c7-82b7-2db12d0bd5b0.png)

**Reply by seikosantana** [reply]
So some notes is i shouldn't be sending responses with the `HttpContext` or else the socket will close.
Seems like by default if no response is sent from a route, it sents a 500 Unimplemented and that causes the socket to on the client.

So for now I have to make it doesn't leave the route action method so it doesn't sent a respond at the end of route invocation.

That doesn't seem to be much problem

**Reply by scottoffen** [reply]
There is a work around to that. Set the [RequireRouteResponse](https://github.com/scottoffen/grapevine/blob/main/src/Grapeseed/RouterOptions.cs#L17) property to false in your startup class.

```csharp
public void ConfigureServer(IRestServer server)
{
    server.Router.Options.RequireRouteResponse = false;
}
```

If not response is sent, the router and server will not see this as an error condition.
